pkg/registry/core/service/ipallocator/controller/repair.go

```diff
@@ -158,6 +157,25 @@ func (c *Repair) runOnce() error {
 		return fmt.Errorf("unable to refresh the service IP block: %v", err)
 	}
 
+	// before we go through snapshots, we fix "PreferDualStack" services
+	// if needed
+	if err = c.fixPreferDualStackServices(); err != nil {
+		return err
+	}
+	// by now we know that snapshots are ready because we poll-ed for them
+	// early on
+	for family, allocator := range c.allocatorByFamily {
+		// get snapshot if it is not there
+		if _, ok := snapshotByFamily[family]; !ok {
+			snapshot, err := allocator.Get()
```

Comment:
just dumping my thoughts here about patching services in this loop:

1. we get the bitmap/allocator here
2. it process it and repair if necessary
3. save it in etcd L300

so far so good, with this, if we have a service with preferDual in an upgrade/downgrade we need to add or remove a ClusterIP.
1. we send a patch
2. this patchs goes to the rest 
3. I can't remember the code path exactly, but this will allocate or release an IP, hence modify the allocator

the repair saveSnapshot will always conflict because we modified the allocator with the patch and the repair loop is working with an older version ... however, the next run, the repair will work because we don't have to patch again the service

Reply:
The sequence is slightly different:
1. Get - but do not save - allocators to ensure that the system is healthy and is ready. This is L145.
2. fix all `PreferDualstack` services if any L162.
3. Get allocators, repair services L167.

Step 2 runs end-to-end before step3 there is no conflicts.

Reply:
ah oh yeah, ok, this Get is after the `fixPreferDual` sorry